libsolidity/formal/CHC.cpp

```diff
@@ -112,48 +112,76 @@ vector<string> CHC::unhandledQueries() const
 bool CHC::visit(ContractDefinition const& _contract)
 {
 	resetContractAnalysis();
-
 	initContract(_contract);
-
-	m_stateVariables = SMTEncoder::stateVariablesIncludingInheritedAndPrivate(_contract);
-
 	clearIndices(&_contract);
 
+	m_stateVariables = SMTEncoder::stateVariablesIncludingInheritedAndPrivate(_contract);
 	solAssert(m_currentContract, "");
-	m_constructorSummaryPredicate = createSymbolicBlock(
-		constructorSort(*m_currentContract, state()),
-		"summary_constructor_" + contractSuffix(_contract),
-		PredicateType::ConstructorSummary,
-		&_contract
-	);
 
 	SMTEncoder::visit(_contract);
 	return false;
 }
 
 void CHC::endVisit(ContractDefinition const& _contract)
 {
-	auto implicitConstructorPredicate = createSymbolicBlock(
-		implicitConstructorSort(state()),
-		"implicit_constructor_" + contractSuffix(_contract),
-		PredicateType::ImplicitConstructor,
-		&_contract
-	);
-	addRule(
-		(*implicitConstructorPredicate)({0, state().thisAddress(), state().crypto(), state().tx(), state().state()}),
-		implicitConstructorPredicate->functor().name
-	);
-	setCurrentBlock(*implicitConstructorPredicate);
-
 	if (auto constructor = _contract.constructor())
 		constructor->accept(*this);
-	else
-		inlineConstructorHierarchy(_contract);
 
-	connectBlocks(m_currentBlock, summary(_contract));
+	defineContractInitializer(_contract);
+
+	auto const& entry = *createConstructorBlock(_contract, "implicit_constructor_entry");
+
+	// In case constructors use uninitialized state variables,
+	// they need to be zeroed.
+	// This is not part of `initialConstraints` because it's only true here,
+	// at the beginning of the deployment routine.
+	smtutil::Expression zeroes(true);
+	for (auto var: stateVariablesIncludingInheritedAndPrivate(_contract))
+		zeroes = zeroes && currentValue(*var) == smt::zeroValue(var->type());
+	addRule(smtutil::Expression::implies(initialConstraints(_contract) && zeroes, predicate(entry)), entry.functor().name);
+	setCurrentBlock(entry);
+
+	solAssert(!m_errorDest, "");
+	m_errorDest = m_constructorSummaries.at(&_contract);
+	// We need to evaluate the base constructor calls (arguments) from derived -> base
+	auto baseArgs = baseArguments(_contract);
+	for (auto base: _contract.annotation().linearizedBaseContracts)
+	{
+		if (base != &_contract)
+		{
+			m_callGraph[&_contract].insert(base);
+			vector<ASTPointer<Expression>> const& args = baseArgs.count(base) ? baseArgs.at(base) : decltype(args){};
```

Comment:
This might be https://github.com/ethereum/solidity/issues/10532

Reply:
Still clang's fault though :p
It's all local so the ref should be fine

Reply:
https://en.cppreference.com/w/cpp/language/lifetime says "The lifetime of a temporary object may be extended by binding to a const lvalue reference or to an rvalue reference (since C++11), see reference initialization for details. "
It's an rvalue here, so it's new in C++11, maybe clang6 didn't quite catch that yet... although C++11 has been around for a while...

Reply:
By the way this will still always be a copy anyhow, though, I think.

Reply:
And I just confirmed in @agroce's docker image: removing the ref on the LHS fixes it...

Reply:
Why would it be a copy in the true case?

Reply:
For the record: short version of the gitter explanation: the common type of the ternary is ``vector<...>`` (no ref), so the true-case constructs that from the const ref by invoking the copy constructor.